gradle-plugin-testing/src/main/java/com/palantir/gradle/plugintesting/PluginTestingPlugin.java

```diff
@@ -16,13 +16,94 @@
+    /**
+     * Used in tests to pick up the current version of this plugin.
+     */
+    static final String PLUGIN_VERSION_PROPERTY_NAME = "pluginTestingPluginVersion";
 
+    /**
+     * Applies the plugin to the given project.
+     */
     @Override
-    public final void apply(Project project) {
-        project.getTasks().register("plugintestingTask", PluginTestingTask.class);
+    public void apply(Project project) {
+        PluginTestingExtension testUtilsExt =
+                project.getExtensions().create(PluginTestingExtension.EXTENSION_NAME, PluginTestingExtension.class);
+
+        addTestDependency(project);
+
+        SourceSetContainer sourceSetContainer = project.getExtensions().getByType(SourceSetContainer.class);
+        SourceSet sourceSet = sourceSetContainer.getByName(SourceSet.TEST_SOURCE_SET_NAME);
+        NamedDomainObjectProvider<Configuration> testRuntimeConfig =
+                project.getConfigurations().named(sourceSet.getRuntimeClasspathConfigurationName());
+
+        project.getTasks().withType(Test.class).configureEach(test -> {
+            // need to use the doFirst so that any custom settings on the extension are applied before reading
+            // the values and setting the system properties.
+            Action<Task> action = new Action<>() {
+                @Override
+                public void execute(Task _task) {
+                    // add system property for all test dependencies so that TestDepVersions can resolve them
+                    Set<String> depSet = getDependencyStrings(testRuntimeConfig.get());
+                    String depsString = String.join(",", depSet);
+                    test.systemProperty(TestDepVersions.TEST_DEPENDENCIES_SYSTEM_PROPERTY, depsString);
+
+                    // add system property for what versions of gradle should be used in tests
+                    String versions =
+                            String.join(",", testUtilsExt.getGradleVersions().get());
+                    test.systemProperty(GradleTestVersions.TEST_GRADLE_VERSIONS_SYSTEM_PROPERTY, versions);
+
+                    // add system property to ignore gradle deprecations so that nebula tests don't fail
+                    if (testUtilsExt.getIgnoreGradleDeprecations().get()) {
+                        // from
+                        // https://github.com/nebula-plugins/nebula-test/blob/main/src/main/groovy/nebula/test/IntegrationBase.groovy
+                        test.systemProperty("ignoreDeprecations", "true");
+                    }
+                }
+            };
+            test.doFirst(action);
+        });
+    }
+
+    /**
+     * Add test dependency on this plugin to the project so can access utility methods from code if needed.  This is
+     * normally done by getting the Implementation_Version setting in the compiled jar, but that doesn't always work
+     * when running tests so we can also look it up via a gradle property that tests set.
```

Comment:
that comment is out of date (I moved the required classes into the "core" module) but the functionality may not be.  It came from Callum's work here - https://github.com/palantir/suppressible-error-prone/blob/bf5fe995006346d5d4bf7345a81530f932cc0ca6/gradle-suppressible-error-prone/src/main/java/com/palantir/gradle/suppressibleerrorprone/SuppressibleErrorPronePlugin.java#L60-L64

I can re check and see if it's still required in this situation.  Since the use case is more simple it may not be.

Reply:
I did re-check that the version information is not found when running tests without the extra logic of passing the current version as a property.  I even tried changing the `PluginTestingPlugin.class.getPackage().getImplementationVersion()` bit to use the TestDependencyVersions class instead because it is (in theory) loaded from a jar in the tests, but it still fails.  I presume this is because gradle somewhere substitutes a project dependency for the jar dependency.  I updated the comment to be more clear though.